genesis/utils/path_planing.py

```diff
@@ -0,0 +1,995 @@
+
+    @abstractmethod
+    def plan(
+        self,
+        qpos_goal,
+        qpos_start=None,
+    ): ...
+
+    def get_link_transformation(self, robot_g_link_idx, obj_g_link_idx):
+        """
+        get transformation from robot link to the object link
+        robot_g_link_idx: int
+            global link idx of the link of the robot
+        obj_g_link_idx: int
+            global link idx of the base link of the object
+        """
+        robot_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=robot_g_link_idx),
+            self._solver.get_links_quat(links_idx=robot_g_link_idx),
+        )
+        obj_T = gg.trans_quat_to_T(
+            self._solver.get_links_pos(links_idx=obj_g_link_idx), self._solver.get_links_quat(links_idx=obj_g_link_idx)
+        )
+        T_robot_object = torch.linalg.inv(robot_T) @ obj_T
+        return T_robot_object
+
+    def update_object(self, ee_link_idx, obj_link_idx, T_robot_object, envs_idx):
+        if self._solver.n_envs > 0:
+            robot_T = gg.trans_quat_to_T(
+                self._solver.get_links_pos(ee_link_idx, envs_idx=envs_idx),
+                self._solver.get_links_quat(ee_link_idx, envs_idx=envs_idx),
+            )
+        else:
+            robot_T = gg.trans_quat_to_T(
+                self._solver.get_links_pos(ee_link_idx), self._solver.get_links_quat(ee_link_idx)
+            )
+        obj_T = robot_T @ T_robot_object
+        trans, quat = gg.T_to_trans_quat(obj_T)
+        if self._solver.n_envs > 0:
+            self._solver.set_base_links_pos(trans, obj_link_idx, envs_idx=envs_idx)
+            self._solver.set_base_links_quat(quat, obj_link_idx, envs_idx=envs_idx)
+        else:
+            self._solver.set_base_links_pos(trans, obj_link_idx)
+            self._solver.set_base_links_quat(quat, obj_link_idx)
+
+    # ------------------------------------------------------------------------------------
+    # ------------------------------ util funcs ------------------------------------------
+    # ------------------------------------------------------------------------------------
+    def format_input(self, qpos_cur, qpos_goal, qpos_start, envs_idx):
+        if qpos_start is None:
+            qpos_start = qpos_cur.clone()
+            if self._solver.n_envs == 0:
+                qpos_start = qpos_start.unsqueeze(0)
+        else:
+            if qpos_start.ndim == 1:
+                qpos_start = qpos_start.unsqueeze(0)
+            assert qpos_start.ndim == 2
+            if qpos_start.shape[0] == 1:
+                qpos_start = qpos_start.repeat(len(envs_idx), 1)
+
+        if qpos_goal.ndim == 1:
+            qpos_goal = qpos_goal.unsqueeze(0)
+        assert qpos_goal.ndim == 2
+        if qpos_goal.shape[0] == 1:
+            qpos_goal = qpos_goal.repeat(len(envs_idx), 1)
+        return qpos_cur, qpos_goal, qpos_start, envs_idx
```

Comment:
You should probably check that the shape of all the tensors are valid before returning.

Reply:
At some point, you had `gs.raise_exception("Invalid shape for 'qpos_start' or 'qpos_goal'.")` but it has been removed. I'm afraid the exception may not be clear enough for the user without checking for size explicitly.